database-commons/src/main/java/io/cdap/plugin/util/RetryUtils.java

```diff
@@ -0,0 +1,141 @@
+      String query, String externalDocumentationLink) {
+        try {
+            return Failsafe.with(retryPolicy).get(() -> statement.executeQuery(query));
+        } catch (Exception e) {
+            throw unwrapFailsafeException(e, externalDocumentationLink);
+        }
+    }
+
+    public static void executeInitQueryWithRetry(RetryPolicy<?> retryPolicy, Statement statement, String query,
+      String externalDocumentationLink) {
+        try {
+            Failsafe.with(retryPolicy).run(() -> statement.execute(query));
+        } catch (Exception e) {
+            throw unwrapFailsafeException(e, externalDocumentationLink);
+        }
+    }
+
+    private static RuntimeException unwrapFailsafeException(Exception e, String externalDocumentationLink) {
+        if (e instanceof FailsafeException) {
+            Throwable cause = e.getCause();
+            if (cause instanceof SQLException) {
+                return programFailureException((SQLException) cause, externalDocumentationLink);
+            } else if (cause instanceof RuntimeException) {
+                return (RuntimeException) cause;
+            } else if (cause instanceof Error) {
+                return new RuntimeException("Failsafe wrapped an Error", cause);
+            } else {
+                return new RuntimeException("Failsafe wrapped a non-runtime exception", cause);
+            }
+        }
+        if (e instanceof SQLException) {
+            return programFailureException((SQLException) e, externalDocumentationLink);
+        }
+        if (e instanceof RuntimeException) {
+            return (RuntimeException) e;
+        }
+        return new RuntimeException("Unexpected checked exception", e);
+    }
+
+    private static ProgramFailureException programFailureException(SQLException e, String externalDocumentationLink) {
+        // wrap exception to ensure SQLException-child instances not exposed to contexts without jdbc
+        // driver in classpath
+        String errorMessage =
+          String.format("SQL Exception occurred: [Message='%s', SQLState='%s', ErrorCode='%s'].", e.getMessage(),
+            e.getSQLState(), e.getErrorCode());
+        String errorMessageWithDetails = String.format("Error occurred while trying to" +
+          " get schema from database." + "Error message: '%s'. Error code: '%s'. SQLState: '%s'", e.getMessage(),
+            e.getErrorCode(), e.getSQLState());
+
+        if (!Strings.isNullOrEmpty(externalDocumentationLink)) {
+            if (!errorMessage.endsWith(".")) {
+                errorMessage = errorMessage + ".";
+            }
+            errorMessage = String.format("%s For more details, see %s", errorMessage, externalDocumentationLink);
+        }
+        return ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN),
+          errorMessage, errorMessageWithDetails, ErrorType.USER, false, ErrorCodeType.SQLSTATE, e.getSQLState(),
```

Comment:
this cannot be `USER` always, we should check based on `SQLState` and dependency should be `true`. Not sure how was this missed earlier.

Reply:
you can use `DBErrorDetailsProvider#getException#getProgramFailureException` : https://github.com/cdapio/hydrator-plugins/blob/b9723d1b24068347c60ad2b0489064902f335de9/hydrator-common/src/main/java/io/cdap/plugin/common/db/DBErrorDetailsProvider.java#L184

Reply:
If we directly use this method, then we wouldn't be able to pass the `externalDocumentationLink` as an argument in the method call to `DBErrorDetailsProvider#getException#getProgramFailureException` unless we create another overloaded method in the `DBErrorDetailsProvider` class to accept this new parameter.

Reply:
Directly calling this method now `DBErrorDetailsProvider#getException#getProgramFailureException` with the implementation done as per our offline conversation.